pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go

```diff
@@ -0,0 +1,140 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +k8s:openapi-gen=true
+// TidbClusterAutoScalerList is TidbClusterAutoScaler list
+type TidbClusterAutoScalerList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +k8s:openapi-gen=false
+	metav1.ListMeta `json:"metadata"`
+
+	Items []TidbClusterAutoScaler `json:"items"`
+}
+
+// +k8s:openapi-gen=true
+// TidbAutoScalerSpec describes the state of the TidbClusterAutoScaler
+type TidbClusterAutoScalerSpec struct {
+	// TidbClusterRef describe the target TidbCluster
+	Cluster TidbClusterRef `json:"cluster"`
+
+	// We used prometheus to fetch the metrics resources until the pd could provide it.
+	// MetricsUrl represents the url to fetch the metrics info
+	// +optional
+	MetricsUrl *string `json:"metricsUrl,omitempty"`
+
+	// Tikv represents the auto-scaling spec for tikv
+	Tikv TikvAutoScalerSpec `json:"tikv"`
+
+	// Tidb represents the auto-scaling spec for tidb
+	Tidb TidbAutoScalerSpec `json:"tidb"`
+}
+
+// +k8s:openapi-gen=true
+// TikvAutoScalerSpec describes the spec for tikv auto-scaling
+type TikvAutoScalerSpec struct {
+	// maxReplicas is the upper limit for the number of replicas to which the autoscaler can scale out.
+	// It cannot be less than minReplicas.
+	MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,3,opt,name=maxReplicas"`
+
+	// minReplicas is the lower limit for the number of replicas to which the autoscaler
+	// can scale down.  It defaults to 1 pod. Scaling is active as long as at least one metric value is
+	// available.
+	// +optional
+	MinReplicas *int32 `json:"minReplicas,omitempty" protobuf:"varint,2,opt,name=minReplicas"`
+
+	// ScaleInIntervalSeconds represents the duration seconds between each auto-scaling-in
+	// If not set, the default ScaleInIntervalSeconds will be set to 500
+	// +optional
+	ScaleInIntervalSeconds *int32 `json:"scaleInIntervalSeconds,omitempty"`
+
+	// ScaleOutIntervalSeconds represents the duration seconds between each auto-scaling-out
+	// If not set, the default ScaleOutIntervalSeconds will be set to 300
+	// +optional
+	ScaleOutIntervalSeconds *int32 `json:"scaleoutIntervalSeconds,omitempty"`
+
+	// metrics contains the specifications for which to use to calculate the
+	// desired replica count (the maximum replica count across all metrics will
+	// be used).  The desired replica count is calculated multiplying the
+	// ratio between the target value and the current value by the current
+	// number of pods.  Ergo, metrics used must decrease as the pod count is
+	// increased, and vice-versa.  See the individual metric source types for
+	// more information about how each type of metric must respond.
+	// If not set, the default metric will be set to 80% average CPU utilization.
+	// +optional
+	Metrics []v2beta2.MetricSpec `json:"metrics,omitempty" protobuf:"bytes,4,rep,name=metrics"`
+}
+
+// +k8s:openapi-gen=true
+// TidbAutoScalerSpec describes the spec for tidb auto-scaling
+type TidbAutoScalerSpec struct {
```

Comment:
`TidbAutoScalerSpec ` is same with `TikvAutoScalerSpec `?

Reply:
Yes, in basic configuration, `TidbAutoScalerSpec` is the same as  `TikvAutoScalerSpec`.

Reply:
Then what about:
```
type baseSpec {
xxx
}
type TidbAutoScalerSpec struct {
baseSpec
}
```

Reply:
updated.